rp2040-hal-macros/src/lib.rs

```diff
@@ -0,0 +1,59 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use quote::quote;
+use syn::{parse, parse_macro_input, Item, ItemFn, Stmt};
+
+#[proc_macro_attribute]
+pub fn entry(args: TokenStream, input: TokenStream) -> TokenStream {
+    let mut f = parse_macro_input!(input as ItemFn);
+
+    if !args.is_empty() {
+        return parse::Error::new(Span::call_site(), "This attribute accepts no arguments")
+            .to_compile_error()
+            .into();
+    }
+
+    let clear_locks: TokenStream = quote!(unsafe {
+        const SIO_BASE: u32 = 0xd0000000;
+        const SPINLOCK0: u32 = SIO_BASE + 0x100;
+        const SPINLOCK_COUNT: u32 = 32;
+        for i in 0..SPINLOCK_COUNT {
+            ::core::ptr::write_volatile((SPINLOCK0 + 4 * i) as *mut u32, 1);
+        }
```

Comment:
I think it would be more idiomatic to make a pointer to the first spinlock, use pointer arithmetic to calculate the offset to the current spinlock, and use write_volatile as an associated function for pointer (instead of passing it in as an argument)

```suggestion
        const SIO_BASE: u32 = 0xd0000000;
        const SPINLOCK0_PTR: *mut u32 = (SIO_BASE + 0x100) as *mut u32;
        const SPINLOCK_COUNT: usize = 32;
        for i in 0..SPINLOCK_COUNT {
            SPINLOCK0_PTR.wrapping_add(i).write_volatile(1);
        }
```